dxr/config.py

```diff
@@ -16,215 +24,270 @@
+class Config(DotSection):
+    """Validation and encapsulation for the DXR config file
+
+    Examples::
 
-        # Test for conflicting plugins settings
-        conflicts = [p for p in self.disabled_plugins if p in self.enabled_plugins]
-        if conflicts:
-            msg = "Plugin: '%s' is both enabled and disabled in '%s'"
-            for p in conflicts:
-                print >> sys.stderr, msg % (p, name)
-            sys.exit(1)
+        # Settings from the [DXR] section:
+        >>> Config(...).target_folder
 
-        # Warn if $jobs isn't used...
-        if "$jobs" not in self.build_command:
-            msg = "Warning: $jobs is not used in build_command for '%s'"
-            print >> sys.stderr, msg % name
+        # Settings from individual trees:
+        >>> Config(...).trees['some-tree'].build_command
 
+        # Settings from plugin-specific sections of trees:
+        >>> Config(...).trees['some-tree'].buglink.url
 
-class ConfigError(Exception):
-    """Some kind of user error in the configuration"""
+    """
+    # Design decisions:
+    # * Keep the explicit [DXR] section name because, otherwise, global options
+    #   can collide with tree names.
+    # * Keep tree.config shortcut. It makes a lot of things shorter and doesn't
+    #   hurt anything, since it's all read-only anyway.
+    # * Crosswire __getattr__ with __getitem__. It makes callers more readable.
+    # * Keep whitespace-delimited lists in config file. I prefer them to
+    #   commas, people will have to change less, and it wasn't a big deal to
+    #   customize.
+    # * Use configobj to parse the ini file and then schema to validate it.
+    #   configobj's favored "validate" module makes it very hard to have
+    #   whitespace-delimited lists and doesn't compose well since its
+    #   validators are strings. The downside is we only report 1 error at a
+    #   time, but this can and should be fixed in the schema lib.
+
+    def __init__(self, input):
+        """Pull in and validate a config file.
+
+        :arg input: A string or dict from which to populate the config
+
+        Raise ConfigError if the configuration is invalid.
+
+        """
+        schema = Schema({
+            'DXR': {
+                'target_folder': AbsPath,
+                Optional('temp_folder', default='/tmp/dxr-temp'): AbsPath,
+                Optional('default_tree', default=None): basestring,
+                Optional('disabled_plugins', default=plugin_list('')): Plugins,
+                Optional('enabled_plugins', default=plugin_list('*')): Plugins,
+                Optional('generated_date',
+                         default=datetime.utcnow()
+                                         .strftime("%a, %d %b %Y %H:%M:%S +0000")):
+                    basestring,
+                Optional('log_folder', default=None): AbsPath,
+                Optional('workers', default=cpu_count()):
+                    Use(int, error='"workers" must be an integer >= 1.'),
```

Comment:
Then they get the late failure they deserve. ;-) Okay, I'll patch this up as well.